pkg/config/shared.go

```diff
@@ -179,99 +188,164 @@ type BlobConfigAWS struct {
+
 // BlobConfigGCP is the blob store configuration for the GCP provider.
 type BlobConfigGCP struct {
 	CredentialsFile string `name:"credentials-file" description:"Path to the GCP credentials JSON file"`
 	Credentials     string `name:"credentials" description:"JSON data of the GCP credentials, if not using JSON file"`
 }
 
-// Blob store configuration.
-type Blob struct {
+// BlobConfig is the blob store configuration.
+type BlobConfig struct {
 	Provider string          `name:"provider" description:"Blob store provider (local|aws|gcp)"`
 	Local    BlobConfigLocal `name:"local"`
 	AWS      BlobConfigAWS   `name:"aws"`
 	GCP      BlobConfigGCP   `name:"gcp"`
 }
 
+// Bucket returns the requested blob bucket using the config.
+func (c BlobConfig) Bucket(ctx context.Context, bucket string) (*blob.Bucket, error) {
+	switch c.Provider {
+	case "local":
+		return ttnblob.Local(ctx, bucket, c.Local.Directory)
+	case "aws":
+		return ttnblob.AWS(ctx, bucket, &aws.Config{
+			Endpoint:    &c.AWS.Endpoint,
+			Region:      &c.AWS.Region,
+			Credentials: credentials.NewCredentials(blobConfigAWSCredentials(c.AWS)),
+		})
+	case "gcp":
+		var jsonCreds []byte
+		if c.GCP.Credentials != "" {
+			jsonCreds = []byte(c.GCP.Credentials)
+		} else if c.GCP.CredentialsFile != "" {
+			var err error
+			jsonCreds, err = ioutil.ReadFile(c.GCP.CredentialsFile)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, errMissingBlobConfig
+		}
+		return ttnblob.GCP(ctx, bucket, jsonCreds)
+	default:
+		return nil, errUnknownBlobProvider.WithAttributes("provider", c.Provider)
+	}
+}
+
+type BlobPathConfig struct {
+	Bucket string `name:"bucket" description:"Bucket to use"`
+	Path   string `name:"path" description:"Path to use"`
+}
+
+func (c BlobPathConfig) IsZero() bool {
+	return c == BlobPathConfig{}
+}
+
 // FrequencyPlansConfig contains the source of the frequency plans.
 type FrequencyPlansConfig struct {
 	Static    map[string][]byte `name:"-"`
 	Directory string            `name:"directory" description:"Retrieve the frequency plans from the filesystem"`
 	URL       string            `name:"url" description:"Retrieve the frequency plans from a web server"`
+	Blob      BlobPathConfig    `name:"blob" description:"Retrieve the frequency plans from a blob"`
 }
 
-// Store returns a frequencyplan.Store fwith a fetcher based on the configuration.
-// The order of precedence is Static, Directory and URL.
-// If neither Static, Directory nor a URL is set, this method returns nil, nil.
-func (c FrequencyPlansConfig) Store() (*frequencyplans.Store, error) {
-	var fetcher fetch.Interface
+// Fetcher returns a fetch.Interface based on the configuration.
+// The order of precedence is Static, Directory, URL and Blob.
+// If neither Static, Directory, URL nor a Blob is set, this method returns nil, nil.
+func (c FrequencyPlansConfig) Fetcher(ctx context.Context, blobConf BlobConfig) (fetch.Interface, error) {
```

Comment:
I'm okay with merging this now, since we don't have a compatibility promise on our internal Go packages, but I still think it's weird to pass a blob config when in most cases (the default) we're not using blob.

Reply:
It's not pretty, but we need something working now.
I think that the whole `pkg/config` approach to solving our problem is incorrect and these kind of ugly workarounds are necessary due to that.